### PR TITLE
Fix SEARCH-411: double commit [2/2]

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchDataStore.cpp [hunk 1]
```diff
@@ -67,15 +68,15 @@ namespace {
 
 class IResearchFlushSubscription final : public FlushSubscription {
  public:
-  explicit IResearchFlushSubscription(TRI_voc_tick_t tick = 0) noexcept
+  explicit IResearchFlushSubscription(uint64_t tick = 0) noexcept
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:71]: Why `tick` type is changed?

File: arangod/IResearch/IResearchDataStore.cpp [hunk 2]
```diff
@@ -550,53 +589,152 @@ void ConsolidationTask::operator()() {
...
-      _avgCleanupTimeMs{nullptr},
-      _consolidationTimeNum{0},
-      _avgConsolidationTimeMs{nullptr},
-      _metricStats{nullptr} {
+      _id(iid) {
   // initialize transaction callback
-  _trxCallback = [this](transaction::Methods& trx,
-                        transaction::Status status) -> void {
+
+  _trxBeforeFinish = [this](transaction::Methods& trx) {
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:602]: NIT
Technically this is "PreCommit"  as transaction may be finished with Abort an this callback would not be called in such case.
```suggestion
  _trxPreCommit = [this](transaction::Methods& trx) {
```

File: arangod/IResearch/IResearchDataStore.cpp [hunk 3]
```diff
@@ -550,53 +589,152 @@ void ConsolidationTask::operator()() {
...
+      }
+    }
+    else {
+      ctx._ctx.AddToFlush();
+    }
+#else
+    ctx._ctx.AddToFlush();
+#endif
+  };
+  _trxAfterFinish = [this](transaction::Methods& trx,
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:669]: NIT This is called on every status change. Not only on commit.
```suggestion
  _trxStatusChange = [this](transaction::Methods& trx,
```

File: arangod/IResearch/IResearchDataStore.cpp [hunk 4]
```diff
@@ -1516,6 +1709,37 @@ Result IResearchDataStore::remove(transaction::Methods& trx,
 
   return {};
 }
+
+bool IResearchDataStore::exists(IResearchDataStore::Snapshot const& snapshot,
+                                LocalDocumentId documentId, bool nested,
+                                TRI_voc_tick_t const* recoveryTick) const {
+  if (recoveryTick == nullptr) {
+    TRI_ASSERT(false) << "Call exists not in recovery";
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:1717]: Why it is wrong to call this method not in the recovery?

**Review by Dronplane — CHANGES_REQUESTED** [review]

File: arangod/IResearch/IResearchDataStore.cpp (hunk 1 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:71]: It's uint64_t from rocksdb so I think it should be uint64_t, anyway it's just using

File: arangod/IResearch/IResearchDataStore.cpp (hunk 4 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:1717]: Probably nothing, but now we don't have such code

File: arangod/IResearch/IResearchDataStore.cpp (hunk 2 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:602]: Hmm, yes, maybe I need to change it

File: arangod/IResearch/IResearchDataStore.cpp (hunk 3 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:669]: In general probably name same as callback type is good, I agree

File: arangod/IResearch/IResearchDataStore.cpp (hunk 4 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchDataStore.cpp:1717]: Then maybe remove this assert?  Or at least  please leave a comment, why this assert is here

Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:1717]: Ok

File: arangod/IResearch/IResearchDataStore.cpp (hunk 1 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchDataStore.cpp:71]: That will confuse readers. As in base class we have  TRI_voc_tick_t . So I'd prefer to be consistent with typenames. Even if for now these are the same.

Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:71]: The main issue for me rocksdb tick not TRI_vocbase, so I think we should also replace to uint64_t

Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:71]: Internally in iresearch also uint64

Comment (Dronplane) [reply arangod/IResearch/IResearchDataStore.cpp:71]: IMO: On server level we operate the abstraction of engine ticks with type "TRI_voc_tick_t". That is not necessary RocksDB ticks (we may replace storage engine in future).  The only requirement is that underlying engine ticks should fit into this type. And all "FlushFeature" related stuff should operate `TRI_voc_tick_t`.

File: arangod/IResearch/IResearchDataStore.cpp [hunk 5]
```diff
@@ -828,36 +963,79 @@ Result IResearchDataStore::commitUnsafeImpl(
...
+    std::move(stageTwoGuard).Cancel();
+    if (!commitTwo) {
+      LOG_TOPIC("21bda", TRACE, TOPIC)
+          << "Second commit for ArangoSearch index '" << id()
+          << "' is no changes tick " << lastTickBeforeCommitTwo;
+      TRI_ASSERT(_lastCommittedTickOne <= lastTickBeforeCommitTwo);
+      TRI_ASSERT(_lastCommittedTickTwo <= lastTickBeforeCommitTwo);
+      // no changes, can release the latest tick before commit two
+      _lastCommittedTickOne = lastTickBeforeCommitTwo;
+      _lastCommittedTickTwo = lastTickBeforeCommitTwo;
```
Comment (Dronplane) [inline arangod/IResearch/IResearchDataStore.cpp:1031]: Github is fooling me or failpoint for crash after stage two is missing? https://github.com/arangodb/arangodb/blob/a4eb4d658241b53f3ede34f983ba31f74609a00d/arangod/IResearch/IResearchLink.cpp#L1163

**Review by Dronplane — APPROVED** [review]
LGTM

File: arangod/IResearch/IResearchDataStore.cpp (hunk 5 above)
Comment (MBkkt) [reply arangod/IResearch/IResearchDataStore.cpp:1031]: True

**Pull request merged**